src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/consumer/PersistentServiceProviderConsumerStore.java

```diff
@@ -0,0 +1,317 @@
+            if (isBlank(publicKeyEncoded) || isBlank(publicKeyAlgorithm) || isBlank(publicKeyFormat)) {
+                Set<String> missingItems = new HashSet<>(3);
+                if (isBlank(publicKeyEncoded)) {
+                    missingItems.add(PUBLIC_KEY_ENCODED);
+                }
+                if (isBlank(publicKeyAlgorithm)) {
+                    missingItems.add(PUBLIC_KEY_ALGORITHM);
+                }
+                if (isBlank(publicKeyFormat)) {
+                    missingItems.add(PUBLIC_KEY_FORMAT);
+                }
+                log.warning(() -> String.format(
+                        "Failed to unmarshal OAuth consumer: '%s' node is missing valid values for the following " +
+                                "child nodes: [%s]", PUBLIC_KEY, String.join(",", missingItems)));
+                throw new StoreException("Failed to unmarshal OAuth consumer");
+            }
+            KeyFactory keyFactory;
+            try {
+                keyFactory = KeyFactory.getInstance(publicKeyAlgorithm);
+            } catch (NoSuchAlgorithmException e) {
+                log.log(WARNING, "Failed to unmarshal OAuth consumer", e);
+                throw new StoreException("Failed to unmarshal OAuth consumer");
+            }
+            byte[] publicKeyDecoded = Base64.getDecoder().decode(publicKeyEncoded);
+            KeySpec keySpec;
+            if ("X.509".equals(publicKeyFormat)) {
+                keySpec = new X509EncodedKeySpec(publicKeyDecoded);
+            } else if ("PKCS#8".equals(publicKeyFormat)) {
+                keySpec = new PKCS8EncodedKeySpec(publicKeyDecoded);
+            } else {
+                log.warning("Unknown public key format: " + publicKeyFormat);
+                throw new StoreException("Failed to unmarshal OAuth consumer");
+            }
+            PublicKey publicKey;
+            try {
+                publicKey = keyFactory.generatePublic(keySpec);
+            } catch (InvalidKeySpecException e) {
+                log.log(WARNING, "Failed to unmarshal OAuth consumer", e);
+                throw new StoreException("Failed to unmarshal OAuth consumer");
+            }
+            return publicKey;
+        }
+
+        private static void addPublicKeyNode(HierarchicalStreamWriter writer, PublicKey pubKey) {
+            writer.startNode(PUBLIC_KEY);
+            addNode(writer, PUBLIC_KEY_ENCODED, Base64.getEncoder().encodeToString(pubKey.getEncoded()));
+            addNode(writer, PUBLIC_KEY_ALGORITHM, pubKey.getAlgorithm());
+            addNode(writer, PUBLIC_KEY_FORMAT, pubKey.getFormat());
+            writer.endNode();
+        }
+
+        @Override
+        public boolean canConvert(Class type) {
+            return Consumer.class == type;
+        }
+
+        @Override
+        public void marshal(Object source, HierarchicalStreamWriter writer, MarshallingContext context) {
+            if (!(source instanceof Consumer)) {
```

Comment:
Considering (theoretically) any subtype of `Consumer` has to adhere to the _is-a_ principle, it will _at least_ have all the data that a `Consumer` would contain, so we should be able to marshal it to XML the same way we'd do with `Consumer`.
Of course, it will be unmarshalled back to `Consumer` and not the subtype, but so be it.
I don't see need to strictly limit marshaling to `Consumer`.
As an aside, for strict type checks, `==` can be used instead of `equals`:
`Consumer.class == source.getClass()`